File: src/main/java/work/nekow/particledrawing/core/server/ParticleData.java
package work.nekow.particledrawing.core.server;

import work.nekow.particledrawing.api.Color;
import work.nekow.particledrawing.api.ParticleStyle;
import net.minecraft.world.phys.Vec3;

import java.util.UUID;

/**
 * Server-authoritative particle state.
 * Lives on the server thread, mutated by tick logic and API calls.
 */
@SuppressWarnings("unused")
public final class ParticleData {

    private final UUID id;
    private final ParticleStyle style;
    private Vec3 position;
    private Color color;
    private float scale;
    private int lifetime;          // remaining ticks
    private final int maxLifetime; // original lifetime, for progress calcs
    private final UUID groupId;
    private boolean glowing;
    private Vec3 offsetFromPivot;  // offset at creation time, for group transforms

    ParticleData(UUID id, ParticleStyle style, Vec3 position, Color color,
                 float scale, int lifetime, UUID groupId, boolean glowing,
                 Vec3 offsetFromPivot) {
        this.id = id;
        this.style = style;
        this.position = position;
        this.color = color;
        this.scale = scale;
        this.lifetime = lifetime;
        this.maxLifetime = lifetime;
        this.groupId = groupId;
        this.glowing = glowing;
        this.offsetFromPivot = offsetFromPivot != null ? offsetFromPivot : Vec3.ZERO;
    }

    public UUID id() { return id; }
    public ParticleStyle style() { return style; }
    public Vec3 position() { return position; }
    public Color color() { return color; }
    public float scale() { return scale; }
    public int lifetime() { return lifetime; }
    public int maxLifetime() { return maxLifetime; }
    public UUID groupId() { return groupId; }
    public boolean glowing() { return glowing; }
    public Vec3 offsetFromPivot() { return offsetFromPivot; }

    public void setPosition(Vec3 position) { this.position = position; }
    public void setColor(Color color) { this.color = color; }
    public void setScale(float scale) { this.scale = scale; }
    public void setLifetime(int lifetime) { this.lifetime = lifetime; }
    public void setGlowing(boolean glowing) { this.glowing = glowing; }
    public void setOffsetFromPivot(Vec3 offset) { this.offsetFromPivot = offset; }

    public boolean isExpired() {
        return lifetime == 0;
    }

    public int tick() {
        if (lifetime > 0) {
            lifetime--;
        }
        return lifetime;
    }

    public float lifeProgress() {
        if (maxLifetime < 0) return 0f;
        return 1f - (float) lifetime / maxLifetime;
    }

    public ParticleSnapshot toSnapshot() {
        return new ParticleSnapshot(id, style, position, color, scale, glowing);
    }

    /**
     * Lightweight immutable snapshot for visibility checks and comparison.
     */
    public record ParticleSnapshot(
        UUID id,
        ParticleStyle style,
        Vec3 position,
        Color color,
        float scale,
        boolean glowing
    ) {}

    public static ParticleData create(UUID id, ParticleStyle style, Vec3 position,
                                       Color color, float scale, int lifetime,
                                       UUID groupId, boolean glowing, Vec3 offsetFromPivot) {
        return new ParticleData(id, style, position, color, scale, lifetime,
                                groupId, glowing, offsetFromPivot);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ParticleData that)) return false;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public String toString() {
        return "ParticleData{" + id + " " + style + " @ " + position + "}";
    }
}

File: src/main/kotlin/work/nekow/particledrawing/core/server/ParticleData.kt
package work.nekow.particledrawing.core.server

import net.minecraft.world.phys.Vec3
import work.nekow.particledrawing.api.Color
import java.util.UUID

// 粒子运行时数据：位置、颜色、缩放、寿命等。
@Suppress("unused")
class ParticleData(
    val id: UUID,
    private var position: Vec3,
    private var color: Color,
    private var scale: Float,
    private var lifetime: Int,
    val maxLifetime: Int,
    val groupId: UUID?,
    private var glowing: Boolean,
    private var lightLevel: Int,
    private var offsetFromPivot: Vec3 = Vec3.ZERO
) {

    private var velocity: Vec3 = Vec3.ZERO

    fun position(): Vec3 = position
    fun color(): Color = color
    fun scale(): Float = scale
    fun lifetime(): Int = lifetime
    fun glowing(): Boolean = glowing
    fun lightLevel(): Int = lightLevel
    fun offsetFromPivot(): Vec3 = offsetFromPivot
    fun velocity(): Vec3 = velocity

    fun setPosition(position: Vec3) { this.position = position }
    fun setColor(color: Color) { this.color = color }
    fun setScale(scale: Float) { this.scale = scale }
    fun setLifetime(lifetime: Int) { this.lifetime = lifetime }
    fun setGlowing(glowing: Boolean) { this.glowing = glowing }
    fun setLightLevel(lightLevel: Int) { this.lightLevel = lightLevel.coerceIn(0, 15) }
    fun setOffsetFromPivot(offset: Vec3) { this.offsetFromPivot = offset }
    fun setVelocity(velocity: Vec3) { this.velocity = velocity }

    fun isExpired(): Boolean = lifetime == 0

    @Suppress("unused")
    fun tick(): Int {
        if (lifetime > 0) {
            lifetime--
        }
        return lifetime
    }

    fun lifeProgress(): Float {
        if (maxLifetime < 0) return 0f
        return 1f - lifetime.toFloat() / maxLifetime
    }

    fun toSnapshot(): ParticleSnapshot {
        return ParticleSnapshot(id, position, color, scale, glowing, lightLevel)
    }

    data class ParticleSnapshot(
        val id: UUID,
        val position: Vec3,
        val color: Color,
        val scale: Float,
        val glowing: Boolean,
        val lightLevel: Int
    )

    companion object {
        fun create(id: UUID, position: Vec3,
                   color: Color, scale: Float, lifetime: Int,
                   groupId: UUID?, glowing: Boolean, lightLevel: Int,
                   offsetFromPivot: Vec3?): ParticleData {
            return ParticleData(id, position, color, scale, lifetime, lifetime,
                groupId, glowing, lightLevel, offsetFromPivot ?: Vec3.ZERO)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ParticleData) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String = "ParticleData{$id @ $position}"
}
